### PR TITLE
Upgrade cpplint

### PR DESCRIPTION
**Pull request by tobodner**
* Our current version of cpplint is a binary dump from over 3 years ago
  - The delta to the new version is significant
* Consuming cpplint as a git submodule is cleaner and avoids hot fixes that we had to do earlier
* Fix linter errors
   - Go over includes along the way

### CODE REVIEW

**Comment by Bouncner** [comment]
I like the chance but why does the "submoduling" affect the number of hot fixes we needed to apply? (I assume hot fixes means "fixes" to get our code passing?)

**Comment by tobodner** [comment]
> I like the chance but why does the "submoduling" affect the number of hot fixes we needed to apply? (I assume hot fixes means "fixes" to get our code passing?)

I am referring to fixes to the cpplint.py file. Instead of consuming upstream fixes, we did them ourselves.

File: .gitmodules [hunk 1]
```diff
@@ -45,4 +48,4 @@
 	url = https://github.com/hyrise-mp/tpcds-kit.git
 [submodule "third_party/zstd"]
 	path = third_party/zstd
-	url = https://github.com/facebook/zstd.git
+	url = https://github.com/facebook/zstd.git
```
Comment (Bensk1) [inline .gitmodules]: FACEBOOK

File: .gitmodules [hunk 2]
```diff
@@ -7,6 +7,9 @@
 [submodule "third_party/cpp-btree"]
 	path = third_party/cpp-btree
 	url = https://github.com/algorithm-ninja/cpp-btree
+[submodule "third_party/cpplint"]
```
Comment (Bensk1) [inline .gitmodules]: Remember to update the dependency file.

**Comment by mrks** [comment]
Is this still WIP?

**Comment by tobodner** [comment]
> Is this still WIP?

Yes, I have to touch a bunch of files to make the new cpplint happy.

File: .gitmodules (hunk 1 above)
Comment (tobodner) [reply .gitmodules]: Please note that this is not new.

**Comment by tobodner** [comment]
> > Is this still WIP?
> 
> Yes, I have to touch a bunch of files to make the new cpplint happy.

Done.

File: src/lib/storage/encoding_type.hpp [hunk 3]
```diff
@@ -63,7 +62,7 @@ bool encoding_supports_data_type(EncodingType encoding_type, DataType data_type)
 
 struct SegmentEncodingSpec {
   constexpr SegmentEncodingSpec() : encoding_type{EncodingType::Dictionary} {}
-  constexpr SegmentEncodingSpec(EncodingType encoding_type_) : encoding_type{encoding_type_} {}
+  constexpr SegmentEncodingSpec(EncodingType encoding_type_) : encoding_type{encoding_type_} {}  // NOLINT
```
Comment (tobodner) [inline src/lib/storage/encoding_type.hpp]: https://stackoverflow.com/questions/17264067/chosen-constructor-is-explicit-in-copy-initialization-error-with-clang-4-2

File: src/benchmark/file_based_benchmark.cpp [hunk 4]
```diff
@@ -1,16 +1,13 @@
+#include <filesystem>
```
Comment (mrks) [inline src/benchmark/file_based_benchmark.cpp]: One of the advantages of having our own cpplint was that we could add filesystem to the list of c++ includes. Having it at the top smells. Can we NOLINT this?

File: src/benchmark/join_order_benchmark.cpp [hunk 5]
```diff
@@ -1,22 +1,16 @@
-#include <boost/algorithm/string.hpp>
-#include <cxxopts.hpp>
 #include <filesystem>
 
-#include <fstream>
+#include <boost/algorithm/string.hpp>
```
Comment (mrks) [inline src/benchmark/join_order_benchmark.cpp]: dto

File: src/benchmarklib/benchmark_item_run_result.cpp [hunk 6]
```diff
@@ -1,4 +1,4 @@
-#include "benchmark_item_result.hpp"
+#include "benchmark_item_run_result.hpp"
```
Comment (mrks) [inline src/benchmarklib/benchmark_item_run_result.cpp]: whoopsie. Was that a cpplint or an IDE catch?

File: src/benchmarklib/tpch/tpch_benchmark_item_runner.cpp [hunk 7]
```diff
@@ -4,15 +4,16 @@ extern "C" {
...
-#include <boost/algorithm/string/join.hpp>
-#include <boost/algorithm/string/replace.hpp>
-#include <boost/date_time/gregorian/gregorian.hpp>
-#include <boost/date_time/posix_time/posix_time.hpp>
 #include <iomanip>
 #include <numeric>
 #include <random>
 #include <sstream>
 
+#include <boost/algorithm/string/join.hpp>
```
Comment (mrks) [inline src/benchmarklib/tpch/tpch_benchmark_item_runner.cpp]: I'm somewhat getting tired of this include order thing. Including external things before our own things makes sense, but can we get lint to ignore the order of external includes?

File: .gitmodules (hunk 2 above)
Comment (tobodner) [reply .gitmodules]: Yes, thanks for the reminder.

File: src/benchmark/file_based_benchmark.cpp (hunk 4 above)
Comment (tobodner) [reply src/benchmark/file_based_benchmark.cpp]: The vanilla cpplint (now) lists filesystem there as well.

File: src/benchmarklib/benchmark_item_run_result.cpp (hunk 6 above)
Comment (tobodner) [reply src/benchmarklib/benchmark_item_run_result.cpp]: This was found by cpplint.

**Comment by tobodner** [comment]
> What would also be interesting is a quick check if some "NOLINTs" for perfectly fine code are still necessary.

I opened a separate issue for this: https://github.com/hyrise/hyrise/issues/1915

**Pull request merged**